### PR TITLE
implement ThisVid extractor (without PhantomJS requirement)

### LINKED ISSUE #17713: [Site support request] ThisVid
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.09.26*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.09.26**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://thisvid.com/videos/im-back6/']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2018.09.26
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.17134
[debug] exe versions: ffmpeg N-91158-g63c4a4b0d6, ffprobe N-91158-g63c4a4b0d6
[debug] Proxy map: {}
[generic] im-back6: Requesting header
WARNING: Falling back on generic information extractor.
[generic] im-back6: Downloading webpage
[generic] im-back6: Extracting information
ERROR: Unsupported URL: https://thisvid.com/videos/im-back6/
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpcaq35cl9\build\youtube_dl\YoutubeDL.py", line 792, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpcaq35cl9\build\youtube_dl\extractor\common.py", line 507, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpcaq35cl9\build\youtube_dl\extractor\generic.py", line 3298, in _real_extract
youtube_dl.utils.UnsupportedError: Unsupported URL: https://thisvid.com/videos/im-back6/
...
<end of log>
```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://thisvid.com/videos/blatino-foot-flash-42/
- Channel: https://thisvid.com/members/516677/
- Category: https://thisvid.com/categories/gay-feet/
- Playlist: https://thisvid.com/playlist/45051/video/smoke-break2/

Note that **youtube-dl does not support sites dedicated to [copyright infringement](https://github.com/rg3/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free)**. In order for site support request to be accepted all provided example URLs should not violate any copyrights.

---

### Description of your *issue*, suggested solution and other information

Site is not supported. Videos are easy to download manually by using the in-built network log of Chrome.

### PR DESCRIPTION
**Pull request by rigstot**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR obsoletes / closes #20940 and closes #17713. For this extractor I have reverse-engineered how the URLs are generated and recreated the algorithm in python. #20940 chose to outsource this task to phantomJS, which may not be available on every users' system.

### CODE REVIEW

File: youtube_dl/extractor/thisvid.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# coding: utf-8
+from __future__ import unicode_literals
+import re
+
+from .common import InfoExtractor
+
+
+class ThisVidIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?thisvid\.com/(?P<type>videos|embed)/(?P<id>[A-Za-z0-9-]+/?)'
```
[the remark this answers is not in the document]
Comment (rigstot) [reply youtube_dl/extractor/thisvid.py]: I'm fairly certain they aren't. I've found one single video with underscores in the title, and the ID still is `/videos/the-shape-of-my-feet/`.

File: youtube_dl/extractor/thisvid.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            'ext': 'mp4',
+            'title': 'Smiler',
+        }
+    }]
+
+    def _real_extract(self, url):
+        main_id = self._match_id(url)
+        webpage = self._download_webpage(url, main_id)
+
+        title = self._html_search_regex(r'<title>(?:Video: )?(.+?)(?: - ThisVid.com)?</title>', webpage, 'title')
```
[the remark this answers is not in the document]
Comment (rigstot) [reply youtube_dl/extractor/thisvid.py]: Good catch!

File: youtube_dl/extractor/thisvid.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        if (re.match(self._VALID_URL, url).group('type') == "videos"):
+            display_id = main_id
+        else:
+            display_id = self._search_regex(r'<link rel="canonical" href="' + self._VALID_URL + r'">', webpage, 'display_id', fatal=False),
+
+        return {
+            'id': video_id,
+            'display_id': display_id,
+            'title': title,
+            'url': getrealurl(video_url, license_code),
```
[the remark this answers is not in the document]
Comment (rigstot) [reply youtube_dl/extractor/thisvid.py]: I've implemented `thumbnail:`, thanks for the reminder!   
`uploader_*` isn't available on `/embed/` (and doesn't really add much value IMO), so I left that out.

I'm unsure about `ext`--wouldn't it be better to let that be deferred until after the download (and then handled elsewhere)? Also, it looks like your version only has it in the tests (which is what I currently have).

**Comment by rigstot** [comment]
I've updated the PR thanks to feedback from jhwgh. The more controversial change may be a check for the version number of the site's javascript code that emits a `report_warning` when the major version changes. I'm not very familiar with ytdl's innards, so I'd be happy about feedback on [this change](https://github.com/ytdl-org/youtube-dl/blob/96c3d967a24ea904d4f9d29cbd8f48eb76ea381c/youtube_dl/extractor/thisvid.py#L36-L39).

**Comment by jhwgh1968** [comment]
Thanks for the improvements, @rigstot! I will start using this myself!

I think ``report_warning`` is the correct behavior based on @dstftw's previous comments, but I will let him decide in the proper code review.

**Comment by rigstot** [comment]
Apologies for being impatient, but what's the usual time line for a PR to get merged into ytdl?

**Comment by rigstot** [comment]
ping @dstftw: may I please ask for a code review and getting this merged? Thanks!

**Comment by rigstot** [comment]
Verified working with player version 5.0.1; updated major version check.

**Comment by rigstot** [comment]
@remitamine and/or @dstftw: is there anything you need from me to get this merged? Thanks.

**Comment by rigstot** [comment]
@remitamine, @dstftw: re-requesting review; thanks.

**Comment by remitamine** [comment]
@rigstot i don't work on porn extractors, so i would appreciate it if you don't mention me or request my review in this issues/PRs.

**Comment by remitamine** [comment]
@rigstot i don't work on porn extractors, so i would appreciate it if you don't mention me or request my review in this issues/PRs.

**Comment by fusey-may2020** [comment]
Works for me with both public and (with cookies) private videos. Would be good to see this integrated.

**Comment by rigstot** [comment]
apologies, i force-pushed a grafted checkout and now i can't repoen this PR. opened a new one here: https://github.com/ytdl-org/youtube-dl/pull/26060

**Pull request closed without merging**